src/foundry_dev_tools/config/token_provider.py

```diff
@@ -0,0 +1,193 @@
+from foundry_dev_tools.errors.meta import FoundryAPIError
+from foundry_dev_tools.utils.config import entry_point_fdt_token_provider
+
+
+class TokenProvider:
+    """Parent class for all TokenProviders.
+
+    TokenProvider implementations always need to have these properties:
+        host: the foundry host, needs to be implemented
+        token: the token from the token provider, needs to be implemented
+    """
+
+    def __init__(self, host: Host):
+        """The TokenProvider base class.
+
+        Args:
+            host: the foundry host
+        """
+        self.host = host
+
+    @property
+    def token(self):
+        """Returns the token from the provider."""
+        msg = "This is only the base TokenProvider class and does not implement getting a token."
+        raise NotImplementedError(msg)
+
+    def requests_auth_handler(self, r: requests.PreparedRequest) -> requests.PreparedRequest:
+        """Sets bearer authentication header on PreparedRequest object.
+
+        Does not overwrite authorization header if present.
+        """
+        r.headers.setdefault("authorization", f"Bearer {self.token}")
+        return r
+
+
+class JWTTokenProvider(TokenProvider):
+    """Provides Host and Token."""
+
+    def __init__(self, host: Host, jwt: Token) -> None:
+        """Initialize the JWTTokenProvider.
+
+        Args:
+            host: the foundry host
+            jwt: the jwt token
+        """
+        super().__init__(host)
+        self._jwt = jwt
+
+    @cached_property
+    def token(self) -> Token:
+        """Returns the token supplied when creating this Provider."""
+        return self._jwt
+
+
+class CachedTokenProvider(TokenProvider):
+    """Parent class for token providers which get their token dynamically and need caching."""
+
+    __cached: Token | None = None
+    __valid_until: float = -1
+    # time to remove from expiry
+    # e.g. it will request a new token if your token expires in 5 seconds
+    __clock_skew: int = 10
+
+    def invalidate_cache(self):
+        """Invalidates the token cache."""
+        self.__cached = None
+        self.__valid_until = -1
+
+    def _request_token(self) -> tuple[Token, float]:
+        """Requests the token from the dynamic source."""
+        msg = "This needs to be implemented by a class, this is just the meta class."
+        raise NotImplementedError(msg)
+
+    @property
+    def token(self) -> Token:
+        """Returns the token from a dynamic source and caches it."""
+        if not self.__cached or self.__valid_until < time.time() + 10:
+            self.__cached, self.__valid_until = self._request_token()
+        return self.__cached
+
+
+DEFAULT_OAUTH_SCOPES = [
+    "offline_access",
+    "compass:view",
+    "compass:edit",
+    "compass:discover",
+    "api:write-data",
+    "api:read-data",
+    "build2:run-build-using-service",
+]
+
+
+class OAuthTokenProvider(CachedTokenProvider):
```

Comment:
Can you add tests for this class, testing the logical flow. E.g. if client_id && authorization code grant -> call to palantir oauth2 library, if client_id && client_secret && client credentials grant -> make http request etc.